### PR TITLE
Fix tests under GCC and MinGW

### PR DESCRIPTION
**Pull request by Mixaill**
* `cmake`: refactor and add GCC support
* `cpuid`: removed unneeded includes, fix cpuid call for GCC
* `ext`: removed `windows.h` include
* `math3`: fix include filenames casing
* `math3`: added lacking defines for non-Windows platforms
* `math3`: replace `rand()` with `XM_RAND()` which max values is limited with `0x8000` (glibc version generates values up to INT32_MAX, which causes test failures with GCC)
* `math3`, `shmath`: replace functions like `_fpclass` with `std::` alternatives
* `xdsp`: replace `_aligned_malloc` with `XM_ALIGNED_MALLOC`, which selects `_aligned_malloc` on MSVC and `std::aligned_alloc` on other toolsets

Requires https://github.com/microsoft/DirectXMath/pull/86

-----

Tested with
* Ubuntu 20.04
   - GCC 9
   - Clang 9
* Windows
  - MSVS 16.4.5
  - Clang-CL 9.0.0

### CODE REVIEW

**Comment by walbourn** [comment]
Took a fair bit of this in [this commit](https://github.com/walbourn/directxmathtest) to validate the MinGW support. If you want to merge or rebase, we can see what's left.

**Comment by Mixaill** [comment]
@walbourn rebased

File: xdsp/Test.cpp [hunk 1]
```diff
@@ -18,14 +18,16 @@
...
+#include <cmath>
+#include <memory>
+
 #include "XDSP.h"
 
 #include <stdio.h>
 #include <float.h>
 #include <malloc.h>
 
-#include <cmath>
```
Comment (walbourn) [inline xdsp/Test.cpp]: These shouldn't need to be before xdsp.h

File: math3/xmvec.cpp [hunk 2]
```diff
@@ -7,6 +7,7 @@
 // http://go.microsoft.com/fwlink/?LinkID=615560
 //-------------------------------------------------------------------------------------
 
+#include <iterator>
```
Comment (walbourn) [inline math3/xmvec.cpp]: Put this header into math3.h

File: math3/math3.h [hunk 3]
```diff
@@ -270,7 +308,9 @@ struct APIFUNCT
 #include <stdio.h>
 #include <stdlib.h>
 
+#ifdef _WIN32
```
Comment (walbourn) [inline math3/math3.h]: I should just get rid of this header and all uses of ``TCHAR``. Better to use explicit W version of Win32 APIs and wchar_t.

**Comment by Mixaill** [comment]
@walbourn switched to `wchar_t` and retested in MSVC, MinGW, GCC and Clang.

**Comment by Mixaill** [comment]
@walbourn rebased

**Comment by walbourn** [comment]
Finished up the changes you proposed in a slightly different form as well as adding ARM64 linux support in https://github.com/walbourn/directxmathtest/pull/11.

Thanks!

**Pull request closed without merging**